torch/csrc/jit/codegen/cuda/lower_misaligned_vectorization.cpp

```diff
@@ -0,0 +1,536 @@
+#include <torch/csrc/jit/codegen/cuda/lower_misaligned_vectorization.h>
+
+#include <torch/csrc/jit/codegen/cuda/index_compute.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+#include <torch/csrc/jit/codegen/cuda/predicate_compute.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+class TORCH_CUDA_CU_API MisalignedVectorizationModifier {
+ public:
+  void process(const std::vector<kir::Expr*>& exprs) {
+    FUSER_PERF_SCOPE("MisalignedVectorizationModifier::process");
+    // Run through loop nests
+    // Find for-loops with misaligned vectorization domains
+    for (auto* expr : exprs) {
+      handle(expr);
+    }
+  }
+
+  kir::Expr* applyReplacements(kir::Expr* expr) const {
+    auto handle_scope = [this](kir::Scope& scope) {
+      for (size_t i = 0; i < scope.size(); ++i) {
+        scope[i] = applyReplacements(scope[i]);
+      }
+    };
+
+    const auto it = loop_replacement_map_.find(expr);
+    if (it != loop_replacement_map_.end()) {
+      return it->second;
+    } else {
+      if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+        handle_scope(for_loop->body());
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        handle_scope(ite->thenBody());
+        handle_scope(ite->elseBody());
+      }
+      return expr;
+    }
+  }
+
+ private:
+  void handle(kir::Expr* expr) {
+    if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      handle(for_loop);
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      handle(ite);
+    }
+  }
+
+  void handle(kir::ForLoop* fl) {
+    for_loops_structure_.push_back(fl);
+
+    // Make copy of exprs because we replace them inplace in fl
+    const auto exprs_copy = fl->body().exprs();
+
+    if (containsAnyDirectChildMisalignedVectorize(fl)) {
+      auto new_fl = handleMisalignedVectorize(for_loops_structure_, fl);
+      loop_replacement_map_.insert({fl, new_fl});
+    } else {
+      for (auto expr : exprs_copy) {
+        handle(expr);
+      }
+    }
+
+    for_loops_structure_.pop_back();
+  }
+
+  void handle(kir::IfThenElse* ite) {
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+  }
+
+  kir::ForLoop* handleMisalignedVectorize(
```

Comment:
It would be great if this function could be divided into logically separate functions as it's long and does a lot of things. Yes, I know our codebase has many even longer functions ;-)

Reply:
I agree that splitting the function apart is a good idea. I kept the code structure relatively the same to limit the number of changes in this PR. I was planning follow-up PRs to make further changes.